nixos/modules/services/web-apps/gancio.nix

```diff
@@ -0,0 +1,280 @@
+            description = "Gancio log level.";
+            type = types.enum [
+              "debug"
+              "info"
+              "warning"
+              "error"
+            ];
+            default = "info";
+          };
+          # FIXME upstream proper journald logging
+          log_path = mkOption {
+            description = "Directory Gancio logs into";
+            readOnly = true;
+            type = types.str;
+            default = "/var/log/gancio";
+          };
+        };
+      };
+      description = ''
+        Configuration for Gancio, see <https://gancio.org/install/config> for supported values.
+      '';
+    };
+
+    userLocale = mkOption {
+      type = with types; attrsOf (attrsOf (attrsOf str));
+      default = { };
+      example = {
+        en.register.description = "My new registration page description";
+      };
+      description = ''
+        Override default locales within gancio.
+        See [https://framagit.org/les/gancio/tree/master/locales](default languages and locales).
+      '';
+    };
+
+    nginx = mkOption {
+      type = types.submodule (import ../web-servers/nginx/vhost-options.nix { inherit config lib; });
+      default = { };
+      example = {
+        enableACME = true;
+        forceSSL = true;
+      };
+      description = "Extra configuration for the nginx virtual host of gancio.";
+    };
+  };
+
+  config = mkIf cfg.enable {
+    environment.systemPackages = [ cfg.package ];
+
+    users.users.${cfg.user} = {
+      isSystemUser = true;
+      group = cfg.user;
+      home = "/var/lib/gancio";
+    };
+    users.groups.${cfg.user} = { };
+
+    systemd.tmpfiles.settings."10-gancio" =
+      let
+        rules = {
+          mode = "0755";
+          user = cfg.user;
+          group = config.users.users.${cfg.user}.group;
+        };
+      in
+      {
+        "/var/lib/gancio/user_locale".d = rules;
+        "/var/lib/gancio/plugins".d = rules;
+      };
+
+    systemd.services.gancio =
+      let
+        configFile = settingsFormat.generate "gancio-config.json" cfg.settings;
+      in
+      {
+        description = "Gancio server";
+        documentation = [ "https://gancio.org/" ];
+
+        wantedBy = [ "multi-user.target" ];
+        after = [
+          "network.target"
+        ] ++ optional (cfg.settings.db.dialect == "postgres") "postgresql.service";
+
+        environment = {
+          NODE_ENV = "production";
+        };
+
+        preStart = ''
+          # We need this so the gancio executable run by the user finds the right settings.
+          ln -sf ${configFile} config.json
+
+          rm -f user_locale/*
+          ${concatStringsSep "\n" (
+            mapAttrsToList (
+              l: c: "ln -sf ${settingsFormat.generate "gancio-${l}-locale.json" c} user_locale/${l}.json"
+            ) cfg.userLocale
+          )}
+
+          rm -f plugins/*
+          ${concatMapStringsSep "\n" (p: "ln -sf ${p} plugins/") cfg.plugins}
+        '';
```

Comment:
I feel like all of this could be better solved by `systemd.tmpfiles.rules`. Or am I missing something? I'd like to reduce shell scripts to the absolute minimum in modules.

Reply:
Why not .rules?

Reply:
https://github.com/NixOS/nixpkgs/pull/279011#discussion_r1747386019